### PR TITLE
Automatically save changes

### PR DESCRIPTION
**Pull request by mgwalker**
Automatically saves the APD as changes are being made. Waits 300 milliseconds after the last change before saving, to prevent flooding the user's computer with a high volume of network traffic. Prevents sending a save while a save is already in progress, since we can't guarantee what order saves will be processed by the server, but queues the new save and sends it when the current one completes.

### This pull request changes...

- adds a redux middleware to handle firing save actions
- updates patch reducer to be smart about not wiping out changes made while a save is ongoing
- updates save action to debounce and queue

### Needs

- review of the concept and behavior
- consideration of the accessibility of this approach - presumably we don't want to announce every save. Maybe we don't need to announce anything as long as the saves are successful because "saved" is now the default state?
- consideration of what this does for multiple simultaneous users
- updated tests

### This pull request is ready to merge when...

- [ ] Tests have been updated (and all tests are passing)
- [x] This code has been reviewed by someone other than the original author
- [x] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented
- [ ] The change has been documented
- [ ] Changelog is updated as appropriate

### This feature is done when...

- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

File: web/src/reducers/patch.test.js [hunk 1]
```diff
@@ -14,8 +14,16 @@ describe('JSON patch reducer', () => {
     expect(reducer(undefined, {})).toEqual([]);
   });
 
-  it('resets to the initial state when an APD is saved', () => {
-    expect(reducer({}, { type: SAVE_APD_SUCCESS })).toEqual([]);
+  it('removes the already-saved entries when an APD is successfully saved', () => {
+    // Prime the reducer by trigger a save request. The state at this point is
+    // marked for saving and should be removed when the safe is cucessful
```
Comment (hbillings) [inline web/src/reducers/patch.test.js]:
```suggestion
    // marked for saving and should be removed when the safe is successful
```

Comment (hbillings) [reply web/src/reducers/patch.test.js]: although that was quite an impressive typo

File: web/src/saveMiddleware.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import {
+  ADD_APD_ITEM,
+  ADD_APD_YEAR,
+  EDIT_APD,
+  REMOVE_APD_ITEM,
+  REMOVE_APD_YEAR
+} from './actions/editApd';
+import { saveApd } from './actions/apd';
+
+const saveMiddleware = (store, { save = saveApd } = {}) => next => action => {
```
Comment (hbillings) [inline web/src/saveMiddleware.js]: This makes my head hurt.

File: web/src/reducers/patch.test.js (hunk 1 above)
Comment (mgwalker) [reply web/src/reducers/patch.test.js]: What the hell? 🤣

File: web/src/saveMiddleware.js (hunk 2 above)
Comment (mgwalker) [reply web/src/saveMiddleware.js]: Yeah... The way this redux middleware works is funky.

**Pull request merged**